### PR TITLE
refactor(list): simplify list component scss

### LINKED ISSUE #6152: list: checkbox secondary action gives aria-label warning
I get the warning:
ARIA: Attribute " aria-label ", required for accessibility, is missing on node

My element is:
`<md-checkbox aria-label="Follow" class="md-secondary" ng-init="..." ng-model=".." ng-click="..."></md-checkbox>`

inside md-list-item
When I check the html code of the page I see that the parent of the checkbox is a `<bottom>` with the aria-label

### LINKED ISSUE #8890: list: dense list cuts off some letters like 'g'
Notice how 'g's get cutoff (at least latest Chrome and Firefox, Windows 10)

http://codepen.io/anon/pen/OXpQmV

### PR DESCRIPTION
**Pull request by DevVersion**
- Simplifies the List components SCSS
- Fixes Dense Mode for List and introduces super elegant approach
- Make SCSS more modular by having a base and dynamic variables
- Remove unnecessary DOM node level for button wrap.
- No longer wrap proxied elements inside of buttons
- Multiple Secondary Items should align properly and have consistent spacing.
- Dense Mode should not cut characaters on different environments.

@EladBezalel and I had a short meeting and I showed him some of the new approaches.

R: @EladBezalel 

Fixes #6152. Fixes #8890 References #8482.

### CODE REVIEW

File: src/components/list/list-base.scss [hunk 1]
```diff
@@ -0,0 +1,300 @@
+/** External component calculations */
+$md-checkbox-width: 3 * $baseline-grid !default;
+//TODO(devversion): use shared variable from buttons.scss
```
Comment (EladBezalel) [inline src/components/list/list-base.scss]: If it's for future work please file an issue and write it's number on the comment

File: src/components/list/list-base.scss [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+      @include md-reset-button();
+    }
+  }
+}
+
+/**
+ * Mixin to style user specified content inside of the list item.
+ */
+@mixin md-list-item-content($_list-icon-width, $_list-avatar-width, $_list-avatar-padding-dense) {
```
Comment (EladBezalel) [inline src/components/list/list-base.scss]: As we talked, I don't really like the `$_` prefix for parameters.. maybe local variables should have it instead?

File: src/components/list/list.js [hunk 3]
```diff
@@ -244,6 +244,7 @@ function mdListItemDirective($mdAria, $mdConstant, $mdUtil, $timeout) {
           wrapIn('div');
         } else if (!tEl[0].querySelector('md-button:not(.md-secondary):not(.md-exclude)')) {
           tEl.addClass('md-no-proxy');
+          tEl.addClass('md-list-item-content');
```
Comment (EladBezalel) [inline src/components/list/list.js]: Are you sure this class should be added during the compile phase?

if so i'd minify the calls to:
```js
tEl.addClass('md-no-proxy md-list-item-content');
```

**Comment by EladBezalel** [comment]
I really like the approach we should consider future work on dense components to behave the same.

> Sort of related question:
> Are we applying `dense` mode automatically on small screens?

@clshortfuse

File: src/components/list/list.js (hunk 3 above)
Comment (DevVersion) [inline src/components/list/list.js]: Yeah that's intentionally. It allows us to easily apply the custom styles for user content.

File: src/components/list/list-base.scss (hunk 2 above)
Comment (DevVersion) [inline src/components/list/list-base.scss]: Yeah we both agreed on that. My concern is just that it should be clear whether a variable is dynamic or static.

File: src/components/list/list-base.scss (hunk 1 above)
Comment (DevVersion) [inline src/components/list/list-base.scss]: Since it's an issue, which is not depending on that PR, it would make sense to create it.
> See #9196

**Comment by DevVersion** [comment]
@EladBezalel - Yeah as discussed in our talk. I'll create a follow-up PR, which makes the dense helpers more generic and usable for other components. I'll create an issue once this PR is available.

> Are we applying dense mode automatically on small screens?

No we don't.

File: src/components/list/list-base.scss [hunk 4]
```diff
@@ -99,6 +99,8 @@ $md-list-item-subheader-height: 6 * $baseline-grid !default;
     margin-top: $baseline-grid;
     margin-bottom: $baseline-grid;
     border-radius: 50%;
+
+    flex-shrink: 0; // Per https://bugs.webkit.org/show_bug.cgi?id=146020
```
Comment (DevVersion) [inline src/components/list/list-base.scss]: It's resolved in the latest Webkit version, but it's not resolved in the latest Safari version (probably in the Technology Preview).

I tested that on my Mac Book Pro and the bug is still valid.

**Comment by DevVersion** [comment]
@topherfangio Friendly ping for review if you find some time ;)

File: src/components/list/list.spec.js [hunk 5]
```diff
@@ -355,25 +341,69 @@ describe('mdListItem directive', function() {
...
+
+    var itemContent = listItem.children().eq(1);
+    var secondaryContainer = itemContent.children().eq(1);
+
+    expect(secondaryContainer).toHaveClass('md-secondary-container');
+
+    // The secondary container should contain the md-checkbox, without any button as parent.
+    var checkboxItem = secondaryContainer.children()[0];
+
+    expect(checkboxItem.nodeName).toBe('MD-CHECKBOX');
```
Comment (DevVersion) [inline src/components/list/list.spec.js]: This does pass, but I will change it.

**Comment by DevVersion** [comment]
@topherfangio Thanks for the review.

1. Yeah I'm aware of that, and the change is intentional (follows the specs - [Image](refactor/list-component-scss))
2. Thanks 😄 
3. Yeah this is also intentional, because the items with a proxied element no longer exceed the height (provides consistency and follows specs [Image](https://i.gyazo.com/da2995b4880f4b2ffa39d7484965d2ff.png))

File: src/components/list/list.spec.js [hunk 6]
```diff
@@ -355,25 +341,69 @@ describe('mdListItem directive', function() {
...
+        '<p>Hello World</p>' +
+        '<md-checkbox class="md-secondary" ng-click="null" ng-model="isChecked"></md-checkbox>' +
+      '</md-list-item>'
+    );
+
+    var checkboxEl = listItem.find('md-checkbox');
+
+    expect($rootScope.isChecked).toBeFalsy();
+
+    var clickListener = listItem[0].querySelector('div');
```
Comment (DevVersion) [inline src/components/list/list.spec.js]: Yeah I agree with that. Just changed and also added a comment :)

File: src/components/list/list.spec.js [hunk 7]
```diff
@@ -355,25 +341,69 @@ describe('mdListItem directive', function() {
...
+      '<md-list-item ng-click="sayHello()">' +
+        '<p>Hello World</p>' +
+        '<md-checkbox class="md-secondary" ng-click="goWild()"></md-checkbox>' +
+      '</md-list-item>'
+    );
+
+    expect(listItem).toHaveClass('md-button-wrap');
+
+    // The item should contain the button executor and the item content.
+    expect(listItem.children().length).toBe(2);
```
Comment (topherfangio) [inline src/components/list/list.spec.js]: 👍

**Comment by jelbourn** [comment]
@DevVersion can you rebase?

**Comment by DevVersion** [comment]
@jelbourn Done.

**Pull request merged**